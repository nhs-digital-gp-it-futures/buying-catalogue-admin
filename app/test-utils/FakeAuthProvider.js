export class FakeAuthProvider {
  // eslint-disable-next-line no-useless-constructor, no-empty-function
  constructor(fakeLogoutMethod = () => {}) {
    this.fakeLogout = fakeLogoutMethod;
  }

  // eslint-disable-next-line class-methods-use-this
  setup(app) {
    app.use((req, res, next) => {
      if (req.cookies && req.cookies.fakeToken) {
        req.user = JSON.parse(req.cookies.fakeToken);
      }
      req.logout = this.fakeLogout;
      next();
    });
  }

  // eslint-disable-next-line class-methods-use-this, no-unused-vars
  login() {
    return (req, res) => {
      res.redirect('http://identity-server/login');
    };
  }

  // eslint-disable-next-line class-methods-use-this, no-unused-vars
  loginCallback() {
    return (req, res) => {
      res.redirect('/');
    };
  }

  // eslint-disable-next-line class-methods-use-this
  logout() {
    return '/signout-callback-oidc';
  }

  // eslint-disable-next-line class-methods-use-this
  authorise() {
    return (req, res, next) => {
      if (!req.user) {
        this.login()(req, res, next);
      } else if (req.user && req.user.organisation) {
        next();
      } else {
        throw new Error('Not authorised matey');
      }
    };
  }
}
