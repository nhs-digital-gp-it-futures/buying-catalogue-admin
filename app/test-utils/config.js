export const apiLocalhost = 'http://localhost:5100';

export const organisationsApiLocalhost = 'http://localhost:8075';

export const apiPath = '/api/v1';

export const clientLocalhost = 'http://localhost:1234';
