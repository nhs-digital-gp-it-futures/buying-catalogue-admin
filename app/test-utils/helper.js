import cheerio from 'cheerio';
import request from 'supertest';

export const extractInnerText = async (element) => {
  const elementInnerText = await element.innerText;
  return elementInnerText.trim();
};

const extractCsrfToken = (res) => {
  const $ = cheerio.load(res.text);
  return $('[name=_csrf]').val();
};

export const getCsrfTokenFromGet = async (app, getPath, authenticationCookie) => {
  let cookies;
  let csrfToken;

  await request(app)
    .get(getPath)
    .set('Cookie', authenticationCookie)
    .then((getRes) => {
      cookies = getRes.headers['set-cookie'];
      csrfToken = extractCsrfToken(getRes);
    });
  return { cookies, csrfToken };
};

export const setFakeCookie = async (app, getPath) => {
  let cookies;
  await request(app)
    .get(getPath)
    .set('Cookie', ['cookie1=cookie1value'])
    .then((getRes) => {
      cookies = getRes.headers['set-cookie'];
    });

  return { modifiedApp: app, cookies };
};
