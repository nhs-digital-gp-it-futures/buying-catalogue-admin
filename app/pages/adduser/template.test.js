import { componentTester, snapshotTest } from '../../test-utils/componentTester';
import context from './manifest.json';

context.backLinkHref = '/organisations/org1';

const setup = {
  template: {
    path: 'pages/adduser/template.njk',
  },
};

const questionsWithErrors = {
  questions: [
    {
      id: 'firstName',
      mainAdvice: 'First name',
      error: [{ message: 'First name is required, First name is too long' }],
    },
    {
      id: 'lastName',
      mainAdvice: 'Last name',
      error: [{ message: 'Last name is required, Last name is too long' }],
    },
    {
      id: 'phoneNumber',
      mainAdvice: 'Telephone number',
      error: [{ message: 'Telephone is required' }],
    },
    {
      id: 'emailAddress',
      mainAdvice: 'Email address',
      error: [{ message: 'Email address is required, Email address is too long, Email address already exists' }],
    },
  ],
  errors: [
    { text: 'First name is required', href: '#firstname' },
    { text: 'First name is too long', href: '#firstname' },
    { text: 'Last name is required', href: '#lastname' },
    { text: 'Last name is too long', href: '#lastname' },
    { text: 'Email is required', href: '#emailaddress' },
    { text: 'Email address is too long', href: '#emailaddress' },
    { text: 'Email address already exists', href: '#emailaddress' },
    { text: 'Telephone number is required', href: '#phonenumber' },
  ],
};

const contextWithErrors = {
  ...context,
  ...questionsWithErrors,
};

describe('organisations add user page', () => {
  it('the page should render', componentTester(setup, (harness) => {
    harness.request(context, ($) => {
      const snapshot = snapshotTest($, '[data-test-id="main-content"]');
      expect(snapshot).toMatchSnapshot();
    });
  }));

  it('the page should render with errors', componentTester(setup, (harness) => {
    harness.request(contextWithErrors, ($) => {
      const snapshot = snapshotTest($, '[data-test-id="main-content"]');
      expect(snapshot).toMatchSnapshot();
    });
  }));

  it('should render a backLink', componentTester(setup, (harness) => {
    harness.request(context, ($) => {
      const backLink = $('[data-test-id="go-back-link"]');
      expect(backLink.length).toEqual(1);
      expect(backLink.text().trim()).toEqual('Go back');
      expect($(backLink).find('a').attr('href')).toEqual('/organisations/org1');
    });
  }));

  it('should render error summary with correct error text and hrefs if there are errors', componentTester(setup, (harness) => {
    harness.request(contextWithErrors, ($) => {
      const errorSummary = $('[data-test-id="error-summary"]');
      const errorArray = $('[data-test-id="error-summary"] li a');
      expect(errorSummary.length).toEqual(1);
      expect(errorArray.length).toEqual(contextWithErrors.errors.length);
      contextWithErrors.errors.forEach((error, i) => {
        expect(errorArray[i].attribs.href).toEqual(error.href);
        expect(errorArray[i].children[0].data.trim()).toEqual(error.text);
      });
    });
  }));

  it('should render a title', componentTester(setup, (harness) => {
    harness.request(context, ($) => {
      const title = $('h1[data-test-id="add-user-page-title"]');
      expect(title.length).toEqual(1);
      expect(title.text().trim()).toEqual(context.title);
    });
  }));

  it('should render a description', componentTester(setup, (harness) => {
    harness.request(context, ($) => {
      const description = $('[data-test-id="add-user-page-description"]');
      expect(description.length).toEqual(1);
      expect(description.text().trim()).toEqual(context.description);
    });
  }));

  it('should render a organisation name subheading', componentTester(setup, (harness) => {
    harness.request(context, ($) => {
      const subheading = $('h2[data-test-id="org-name-subheading"]');
      expect(subheading.length).toEqual(1);
      expect(subheading.text().trim()).toEqual(context.orgNameSubheading);
    });
  }));

  it('should render the organisation name', componentTester(setup, (harness) => {
    const modifiedContext = { ...context, organisationName: 'a lovely org name' };
    harness.request(modifiedContext, ($) => {
      const organisationName = $('[data-test-id="org-name"]');
      expect(organisationName.length).toEqual(1);
      expect(organisationName.text().trim()).toEqual(modifiedContext.organisationName);
    });
  }));

  describe('form fields', () => {
    it('should render a label for each question', componentTester(setup, (harness) => {
      harness.request(context, ($) => {
        const labels = $('label');
        expect(labels.length).toEqual(context.questions.length);
        context.questions.forEach((question, i) => {
          expect(labels[i].attribs.for).toEqual(question.id);
          expect(labels[i].children[0].data.trim()).toEqual(question.mainAdvice);
        });
      });
    }));

    it('should render a textField for each question', componentTester(setup, (harness) => {
      harness.request(context, ($) => {
        const inputs = $('input:not([name=_csrf])');
        expect(inputs.length).toEqual(context.questions.length);
        context.questions.forEach((question, i) => {
          expect(inputs[i].attribs.id).toEqual(question.id);
          expect(inputs[i].attribs.name).toEqual(question.id);
          expect(inputs[i].attribs.type).toEqual('text');
        });
      });
    }));

    it('should render errors for each question if there are errors', componentTester(setup, (harness) => {
      harness.request(contextWithErrors, ($) => {
        contextWithErrors.questions.forEach((question) => {
          const renderedQuestion = $(`div[data-test-id="question-${question.id}"]`);
          expect(renderedQuestion.find('div[data-test-id="text-field-input-error"]').length).toEqual(1);
          expect(renderedQuestion.find('.nhsuk-error-message').text().trim()).toEqual('Error:');
        });
      });
    }));
  });

  it('should render a add user button', componentTester(setup, (harness) => {
    harness.request(context, ($) => {
      const button = $('[data-test-id="add-user-button"] button');
      expect(button.length).toEqual(1);
      expect(button.text().trim()).toEqual(context.addUserButtonText);
    });
  }));
});
