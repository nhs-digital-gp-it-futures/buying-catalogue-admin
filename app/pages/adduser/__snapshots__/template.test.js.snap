// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organisations add user page the page should render 1`] = `
"<main class=\\"nhsuk-main-wrapper\\" id=\\"maincontent\\" data-test-id=\\"main-content\\">
        
  <div class=\\"nhsuk-width-container\\" data-test-id=\\"add-user-page\\">
    <div data-test-id=\\"go-back-link\\">
      <div class=\\"nhsuk-back-link\\">
  <a class=\\"nhsuk-back-link__link\\" href=\\"/organisations/org1\\">
  <svg class=\\"nhsuk-icon nhsuk-icon__chevron-left\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\">
    <path d=\\"M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z\\"></path>
  </svg>
  Go back</a>
</div>

    </div>
    <div class=\\"nhsuk-grid-row\\">
      <div class=\\"nhsuk-grid-column-two-thirds\\">
        

        <h1 data-test-id=\\"add-user-page-title\\" class=\\"nhsuk-u-margin-top-7 nhsuk-u-margin-bottom-3\\">Add new user</h1>
        <p data-test-id=\\"add-user-page-description\\" class=\\"nhsuk-heading-l nhsuk-u-font-size-24 nhsuk-u-margin-bottom-6\\">Provide the following information to make a new user account.</p>

        <h2 data-test-id=\\"org-name-subheading\\" class=\\"nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\">Organisation name</h2>
        
  


        <form method=\\"post\\">
          <input type=\\"hidden\\" name=\\"_csrf\\" value=\\"\\">
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-firstName\\">
    
      <div data-test-id=\\"text-field-input\\">
        

<div class=\\"nhsuk-form-group nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"firstName\\">
    First name
  </label>
  <input class=\\"nhsuk-input\\" id=\\"firstName\\" name=\\"firstName\\" type=\\"text\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-lastName\\">
    
      <div data-test-id=\\"text-field-input\\">
        

<div class=\\"nhsuk-form-group nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"lastName\\">
    Last name
  </label>
  <input class=\\"nhsuk-input\\" id=\\"lastName\\" name=\\"lastName\\" type=\\"text\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-phoneNumber\\">
    
      <div data-test-id=\\"text-field-input\\">
        

<div class=\\"nhsuk-form-group nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"phoneNumber\\">
    Telephone number
  </label>
  <input class=\\"nhsuk-input\\" id=\\"phoneNumber\\" name=\\"phoneNumber\\" type=\\"text\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-emailAddress\\">
    
      <div data-test-id=\\"text-field-input\\">
        

<div class=\\"nhsuk-form-group nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"emailAddress\\">
    Email address
  </label>
  <input class=\\"nhsuk-input\\" id=\\"emailAddress\\" name=\\"emailAddress\\" type=\\"text\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          

          <div data-test-id=\\"add-user-button\\" class=\\"nhsuk-u-margin-top-9\\">
            


  
    
  




<button class=\\"nhsuk-button nhsuk-u-margin-bottom-9\\" type=\\"submit\\">
  Add user
</button>



          </div>
        </form>
      </div>
    </div>
  </div>

      </main>"
`;

exports[`organisations add user page the page should render with errors 1`] = `
"<main class=\\"nhsuk-main-wrapper\\" id=\\"maincontent\\" data-test-id=\\"main-content\\">
        
  <div class=\\"nhsuk-width-container\\" data-test-id=\\"add-user-page\\">
    <div data-test-id=\\"go-back-link\\">
      <div class=\\"nhsuk-back-link\\">
  <a class=\\"nhsuk-back-link__link\\" href=\\"/organisations/org1\\">
  <svg class=\\"nhsuk-icon nhsuk-icon__chevron-left\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\">
    <path d=\\"M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z\\"></path>
  </svg>
  Go back</a>
</div>

    </div>
    <div class=\\"nhsuk-grid-row\\">
      <div class=\\"nhsuk-grid-column-two-thirds\\">
        
          
  <div data-test-id=\\"error-summary\\">
    <div class=\\"nhsuk-error-summary\\" aria-labelledby=\\"error-summary-title\\" role=\\"alert\\" tabindex=\\"-1\\">
  <h2 class=\\"nhsuk-error-summary__title\\" id=\\"error-summary-title\\">
    There is a problem
  </h2>
  <div class=\\"nhsuk-error-summary__body\\"><ul class=\\"nhsuk-list nhsuk-error-summary__list\\">
      <li>
        <a href=\\"#firstname\\">First name is required</a>
      </li>
      <li>
        <a href=\\"#firstname\\">First name is too long</a>
      </li>
      <li>
        <a href=\\"#lastname\\">Last name is required</a>
      </li>
      <li>
        <a href=\\"#lastname\\">Last name is too long</a>
      </li>
      <li>
        <a href=\\"#emailaddress\\">Email is required</a>
      </li>
      <li>
        <a href=\\"#emailaddress\\">Email address is too long</a>
      </li>
      <li>
        <a href=\\"#emailaddress\\">Email address already exists</a>
      </li>
      <li>
        <a href=\\"#phonenumber\\">Telephone number is required</a>
      </li>
    </ul>
  </div>
</div>

  </div>

        

        <h1 data-test-id=\\"add-user-page-title\\" class=\\"nhsuk-u-margin-top-7 nhsuk-u-margin-bottom-3\\">Add new user</h1>
        <p data-test-id=\\"add-user-page-description\\" class=\\"nhsuk-heading-l nhsuk-u-font-size-24 nhsuk-u-margin-bottom-6\\">Provide the following information to make a new user account.</p>

        <h2 data-test-id=\\"org-name-subheading\\" class=\\"nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\">Organisation name</h2>
        
  


        <form method=\\"post\\">
          <input type=\\"hidden\\" name=\\"_csrf\\" value=\\"\\">
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-firstName\\">
    
      <div data-test-id=\\"text-field-input-error\\">
        

<div class=\\"nhsuk-form-group nhsuk-form-group--error nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"firstName\\">
    First name
  </label>
  <span class=\\"nhsuk-error-message\\" id=\\"firstName-error\\">
    <span class=\\"nhsuk-u-visually-hidden\\">Error:</span> 
  </span>
  <input class=\\"nhsuk-input nhsuk-input--error\\" id=\\"firstName\\" name=\\"firstName\\" type=\\"text\\" aria-describedby=\\"firstName-error\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-lastName\\">
    
      <div data-test-id=\\"text-field-input-error\\">
        

<div class=\\"nhsuk-form-group nhsuk-form-group--error nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"lastName\\">
    Last name
  </label>
  <span class=\\"nhsuk-error-message\\" id=\\"lastName-error\\">
    <span class=\\"nhsuk-u-visually-hidden\\">Error:</span> 
  </span>
  <input class=\\"nhsuk-input nhsuk-input--error\\" id=\\"lastName\\" name=\\"lastName\\" type=\\"text\\" aria-describedby=\\"lastName-error\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-phoneNumber\\">
    
      <div data-test-id=\\"text-field-input-error\\">
        

<div class=\\"nhsuk-form-group nhsuk-form-group--error nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"phoneNumber\\">
    Telephone number
  </label>
  <span class=\\"nhsuk-error-message\\" id=\\"phoneNumber-error\\">
    <span class=\\"nhsuk-u-visually-hidden\\">Error:</span> 
  </span>
  <input class=\\"nhsuk-input nhsuk-input--error\\" id=\\"phoneNumber\\" name=\\"phoneNumber\\" type=\\"text\\" aria-describedby=\\"phoneNumber-error\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          
            <div class=\\"nhsuk-u-margin-bottom-4\\">
              
  
    
  

  

  <div data-test-id=\\"question-emailAddress\\">
    
      <div data-test-id=\\"text-field-input-error\\">
        

<div class=\\"nhsuk-form-group nhsuk-form-group--error nhsuk-u-margin-bottom-2\\">
  <label class=\\"nhsuk-label nhsuk-u-font-size-24 nhsuk-u-font-weight-bold nhsuk-u-margin-bottom-2\\" for=\\"emailAddress\\">
    Email address
  </label>
  <span class=\\"nhsuk-error-message\\" id=\\"emailAddress-error\\">
    <span class=\\"nhsuk-u-visually-hidden\\">Error:</span> 
  </span>
  <input class=\\"nhsuk-input nhsuk-input--error\\" id=\\"emailAddress\\" name=\\"emailAddress\\" type=\\"text\\" aria-describedby=\\"emailAddress-error\\">
</div>

      </div>
    
    <div data-test-id=\\"text-field-footer\\">
      <span class=\\"nhsuk-label bc-t-font-color--grey-1 nhsuk-u-font-size-16\\">
        
      </span>
    </div>
  </div>

            </div>
          

          <div data-test-id=\\"add-user-button\\" class=\\"nhsuk-u-margin-top-9\\">
            


  
    
  




<button class=\\"nhsuk-button nhsuk-u-margin-bottom-9\\" type=\\"submit\\">
  Add user
</button>



          </div>
        </form>
      </div>
    </div>
  </div>

      </main>"
`;
